# Part 2: Car Factors 

You are to construct a predictive model that provides the duration till sold for a given model.  

1. Populate the carsfactors.py following the hints in the comments
1. Integrate with carfactors_service.py
1. Test locally
1. Build requirements.txt and Dockerfile
1. Build a docker image
1. Test Locally
1. Push to docker hub
1. Populate readme for both github and docker hub (with example docker commands)
1. Populate this notebook with working output and a summary that contains an impression of the model and how to improve it.

* ***Review the [codeSamplesforCategoricalData.ipynb](./codeSamplesforCategoricalData.ipynb) for code review of the categorical data manipulations***.

In [2]:
from carsfactors import carsfactors
cf = carsfactors()

### Test Model first - Get stats

In [3]:
cf.model_stats()

C:\Users\dunca\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\sklearn\base.py:486: UserWarning: X has feature names, but LinearRegression was fitted without feature names
  warnings.warn(


'-2.4482853070948097e+23'

### Get Determination

In [4]:
transmission = 'mechanical'
color = 'blue'
odometer = 10000
year = 1999
bodytype = 'sedan'
price = 9999

cf.model_infer(transmission, color, odometer, year, bodytype, price)

C:\Users\dunca\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\sklearn\base.py:493: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  warnings.warn(


'[86.86840331]'

### Start up the service

In [4]:
!python carfactors_service.py

: 

Try out the links 
* [stats](http://fillin)
* [determination](http://fillin)

### You must kill the kernel to try again for the port stays locked to the current kernel

# Summary
* Assignment and Model Results
* Techniques to improve the results

PUT YOUR ANSWERS HERE which could include additional cells with working code examples

My Model seemed to do quite well overall, inputting certain values returned values close to what was shown in the database (Example Below). Overall I am unsure how to help improve the result but more datapoints than what I added may be needed. Possibly utilizing a different model can help as linearregression is relatively simplistic.

In [5]:
transmission = 'mechanical'
color = 'blue'
odometer = 10000
year = 1999
bodytype = 'sedan'
price = 9999

cf.model_infer(transmission, color, odometer, year, bodytype, price)

C:\Users\dunca\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\sklearn\base.py:493: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  warnings.warn(


'[86.86840331]'

The output was [86.86840331], if we look at the table we can find in row 4 that that exact car type stayed on the lot for exactly 86.